File: src/lib/tool-manifest/tools/get-active-selection.ts
import { z } from 'zod';
import { useEditorStore } from '@/store';
import { maskStore } from '@/core/mask-store';
import type { ToolManifest } from '../types';

const input = z.object({}).describe('No input.');

const output = z.object({
  hasSelection: z.boolean(),
  state: z.enum(['active', 'committed', 'none']),
  label: z.string().nullable().describe('Region label if the selection came from a Claude-named region.'),
  width: z.number().optional(),
  height: z.number().optional(),
  source: z.string().optional().describe('Provenance: sam-point, sam-box, ai-proposed, brush, …'),
});

export const getActiveSelectionTool: ToolManifest<typeof input, typeof output> = {
  name: 'get_active_selection',
  kind: 'query',
  description:
    'Inspect the currently armed selection. Use this before calling select_* tools to avoid clobbering a useful selection, or to find out the label of what is currently in scope.',
  inputSchema: input,
  outputSchema: output,
  handler: () => {
    const s = useEditorStore.getState();
    const ref = s.activeMaskRef ?? s.committedMaskRef;
    if (!ref) {
      return { hasSelection: false, state: 'none' as const, label: null };
    }
    const mask = maskStore.get(ref);
    return {
      hasSelection: true,
      state: s.activeMaskRef ? ('active' as const) : ('committed' as const),
      label: mask?.label ?? null,
      width: mask?.width,
      height: mask?.height,
      source: mask?.source,
    };
  },
};

File: backend/app/tools/atomic/get_active_selection.py
from __future__ import annotations

from pydantic import BaseModel

from app.state.document import SessionDocument
from app.tools.base import BackendTool, ToolPermissions


class _Input(BaseModel):
    pass


class _Output(BaseModel):
    has_selection: bool
    state: str
    label: str | None = None
    width: int | None = None
    height: int | None = None
    source: str | None = None


class GetActiveSelectionTool(BackendTool[_Input, _Output]):
    name = "get_active_selection"
    kind = "query"
    description = (
        "Inspect the currently armed selection mask. Use this before select_* "
        "tools to avoid clobbering a useful selection."
    )
    input_schema = _Input
    output_schema = _Output
    permissions = ToolPermissions(requires_image=False)

    async def handler(self, doc: SessionDocument, input: _Input) -> _Output:  # noqa: A002
        mid = doc.active_mask_id or doc.committed_mask_id
        if mid is None or mid not in doc.masks:
            return _Output(has_selection=False, state="none")
        m = doc.masks[mid]
        state = "active" if doc.active_mask_id == mid else "committed"
        return _Output(
            has_selection=True, state=state,
            label=m.label, width=m.width, height=m.height,
            source=m.source,
        )
